common/common.go

```diff
@@ -695,3 +696,62 @@ func SetIfZero[T comparable](p *T, def T) bool {
 	*p = def
 	return true
 }
+
+var (
+	contextKeys   []any
+	contextKeysMu sync.RWMutex
+)
+
+// RegisterContextKey registers a key to be captured by FreezeCtx
+func RegisterContextKey(key any) {
+	contextKeysMu.Lock()
+	defer contextKeysMu.Unlock()
+	if !slices.Contains(contextKeys, key) {
+		contextKeys = append(contextKeys, key)
+	}
+}
+
+// FrozenContext holds captured context values
+type FrozenContext map[any]any
+
+// FreezeCtx captures values from the context for registered keys
+func FreezeCtx(ctx context.Context) FrozenContext {
+	contextKeysMu.RLock()
+	defer contextKeysMu.RUnlock()
+
+	values := make(FrozenContext, len(contextKeys))
+	for _, key := range contextKeys {
+		if val := ctx.Value(key); val != nil {
+			values[key] = val
+		}
+	}
+	return values
+}
+
+// ThawCtx creates a new context from the frozen context using context.Background() as parent
+func ThawCtx(fc FrozenContext) (context.Context, error) {
+	return MergeCtx(context.Background(), fc)
+}
+
+// MergeCtx adds the frozen values to an existing context
+func MergeCtx(ctx context.Context, fc FrozenContext) (context.Context, error) {
+	for k := range fc {
+		if ctx.Value(k) != nil {
+			return nil, fmt.Errorf("%w: %q", errDuplicateContextKey, k)
+		}
+	}
+	return &mergeCtx{Context: ctx, frozen: fc}, nil
+}
+
+// mergeCtx is a context that merges values from a frozen context and a parent context.
+type mergeCtx struct {
+	context.Context //nolint:containedctx // Using context.WithValue will nest contexts and cause lookup latency
+	frozen          FrozenContext
```

Comment:
🚧 Change request:
* This comment has nothing to do with the linter issue
* As we showed, it's the setting that's more painful, not the lookup, so we should write that. Otherwise people will read it, say "no it doesn't and revert".
```suggestion
	context.Context //nolint:containedctx // mergeCtx implements context.Context
	frozen          FrozenContext
```